File: Graphs/shortestAlternatingPaths.js
/*You are given an integer n, the number of nodes in a directed graph where the nodes are labeled from 0 to n - 1. 
Each edge is red or blue in this graph, and there could be self-edges and parallel edges.

You are given two arrays redEdges and blueEdges where:

redEdges[i] = [ai, bi] indicates that there is a directed red edge from node ai to node bi in the graph, and
blueEdges[j] = [uj, vj] indicates that there is a directed blue edge from node uj to node vj in the graph.

Return an array answer of length n, 
where each answer[x] is the length of the shortest path from node 0 to node x 
such that the edge colors alternate along the path, 
or -1 if such a path does not exist. */

class PathState {
    constructor(node, colorRequired) {
        this.node = node;
        this.colorRequired = colorRequired;
    }
}

const shortestAlternatingPaths = (n, redEdges, blueEdges) => {
    const addToGraph = (edges, color) => {
        const colorGraph = graph.get(color);
        for (let node = 0; node < n; node++) {
            colorGraph.set(node, []);
        }

        for (let [nodeA, nodeB] of edges) {
            colorGraph.get(nodeA).push(nodeB);
        }
    }

    const RED = 0;
    const BLUE = 1;

    const graph = new Map();
    graph.set(RED, new Map());
    graph.set(BLUE, new Map());

    addToGraph(redEdges, RED);
    addToGraph(blueEdges, BLUE);

    const seen = new Array(n).fill().map(() => new Array(2).fill(false));
    seen[0][RED] = seen[0][BLUE] = true;
    
    const shortestPaths = new Array(n).fill(-1);

    let queue = [new PathState(0, RED), new PathState(0, BLUE)];
    let distance = 0;

    while (queue.length) {
        const nextQueue = [];

        for (let {node, colorRequired} of queue) {
            if (shortestPaths[node] === -1) {
                shortestPaths[node] = distance;
            }

            const neighbors = graph.get(colorRequired).get(node);
            for (let neighbor of neighbors) {
                const nextColorRequired = 1 - colorRequired;

                if (!seen[neighbor][nextColorRequired]) {
                    seen[neighbor][nextColorRequired] = true;
                    nextQueue.push(new PathState(neighbor, nextColorRequired));
                }
            }
        }

        queue = nextQueue;
        distance++;
    }

    return shortestPaths;
}

const n = 3, redEdges = [[0,1],[1,2]], blueEdges = []
console.log(shortestAlternatingPaths(n, redEdges, blueEdges));

File: Graphs/ShortestAlternatingPaths.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;

/*You are given an integer n, the number of nodes in a directed graph where the nodes are labeled from 0 to n - 1. 
Each edge is red or blue in this graph, and there could be self-edges and parallel edges.

You are given two arrays redEdges and blueEdges where:

redEdges[i] = [ai, bi] indicates that there is a directed red edge from node ai to node bi in the graph, and
blueEdges[j] = [uj, vj] indicates that there is a directed blue edge from node uj to node vj in the graph.

Return an array answer of length n, 
where each answer[x] is the length of the shortest path from node 0 to node x 
such that the edge colors alternate along the path, 
or -1 if such a path does not exist. */

public class ShortestAlternatingPaths {
    int n;
    private static final int RED = 0;
    private static final int BLUE = 1;
    Map<Integer, ArrayList<ArrayList<Integer>>> graph = new HashMap<>();

    public int[] shortestAlternatingPaths(int n, int[][] redEdges, int[][] blueEdges) {
        this.n = n;
        graph.put(RED, new ArrayList<>());
        graph.put(BLUE, new ArrayList<>());

        generateAdjacencyList(RED, redEdges);
        generateAdjacencyList(BLUE, blueEdges);

        int[] shortestPaths = new int[n];
        Arrays.fill(shortestPaths, -1);

        // Queue stores int arrays of the form [node, prev_color, steps]
        Deque<int[]> queue = new ArrayDeque<>();
        queue.add(new int[] { 0, RED, 0 });
        queue.add(new int[] { 0, BLUE, 0 });

        boolean[][] seen = new boolean[n][2];
        seen[0][RED] = seen[0][BLUE] = true;

        while (!queue.isEmpty()) {
            int[] state = queue.pollFirst();
            int node = state[0];
            int prevColor = state[1];
            int steps = state[2];

            if (shortestPaths[node] == -1) {
                shortestPaths[node] = steps;
            }

            int currentColor = 1 - prevColor;
            for (int neighbor : graph.get(currentColor).get(node)) {
                if (!seen[neighbor][currentColor]) {
                    seen[neighbor][currentColor] = true;
                    queue.addLast(new int[] { neighbor, currentColor, steps + 1 });
                }
            }
        }

        return shortestPaths;
    }

    public void generateAdjacencyList(int color, int[][] edges) {
        ArrayList<ArrayList<Integer>> adjacencyList = graph.get(color);
        for (int node = 0; node < n; node++) {
            adjacencyList.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            adjacencyList.get(edge[0]).add(edge[1]);
        }
    }
}

// Time Complexity: O(n + r + b) where n is the number of nodes,
// r is the number of red edges,
// and b is the number of blue edges.
// Space Complexity: O(n + r + b) for the graph representation and the seen
// array.